src/query/src/range_select/plan_rewrite.rs

```diff
@@ -0,0 +1,466 @@
+    }
+}
+
+pub struct RangePlanRewriter {
+    table_provider: DfTableSourceProvider,
+    context_provider: DfContextProviderAdapter,
+}
+
+impl RangePlanRewriter {
+    pub fn new(
+        table_provider: DfTableSourceProvider,
+        context_provider: DfContextProviderAdapter,
+    ) -> Self {
+        Self {
+            table_provider,
+            context_provider,
+        }
+    }
+
+    pub async fn rewrite(&mut self, plan: LogicalPlan) -> Result<LogicalPlan> {
+        match self.rewrite_logical_plan(&plan).await? {
+            Some(new_plan) => Ok(new_plan),
+            None => Ok(plan),
+        }
+    }
+
+    #[async_recursion]
+    async fn rewrite_logical_plan(&mut self, plan: &LogicalPlan) -> Result<Option<LogicalPlan>> {
+        let inputs = plan.inputs();
+        let mut new_inputs = Vec::with_capacity(inputs.len());
+        for input in &inputs {
+            new_inputs.push(self.rewrite_logical_plan(input).await?)
+        }
+        match plan {
+            LogicalPlan::Projection(Projection { expr, input, .. })
+                if have_range_in_exprs(expr) =>
+            {
+                let input = if let Some(new_input) = new_inputs[0].take() {
+                    Arc::new(new_input)
+                } else {
+                    input.clone()
+                };
+                let (time_index, default_by) = self.get_index_by(input.schema().clone()).await?;
+                let mut range_rewriter = RangeRewriter {
+                    align: Duration::default(),
+                    by: vec![],
+                    range_fn: vec![],
+                    context_provider: &self.context_provider,
+                };
+                let new_expr = expr
+                    .iter()
+                    .map(|expr| expr.clone().rewrite(&mut range_rewriter))
+                    .collect::<DFResult<Vec<_>>>()
+                    .context(DataFusionSnafu)?;
+                if range_rewriter.by.is_empty() {
+                    range_rewriter.by = default_by;
+                }
+                let range_plan = LogicalPlan::Extension(Extension {
+                    node: Arc::new(RangeSelect::try_new(
+                        input.clone(),
+                        range_rewriter.range_fn,
+                        range_rewriter.align,
+                        time_index,
+                        range_rewriter.by,
+                    )?),
+                });
+                // If the result of the project plan happens to be the schema of the range plan, no project plan is required
+                // that need project is identical to range plan schema
+                let need_project = new_expr.iter().any(|expr| {
+                    if let Expr::Column(column) = expr {
+                        !range_plan.schema().has_column(column)
+                    } else {
+                        true
+                    }
+                }) || new_expr.len() != range_plan.schema().fields().len();
```

Comment:
Could you please assign the result of `iter() + any()` to a named variable that explains what it does? The bool expression is hard to understand.
```rust
                let xxx = new_expr.iter().any(|expr| {
                    if let Expr::Column(column) = expr {
                        !range_plan.schema().has_column(column)
                    } else {
                        true
                    }
                });
                let need_project = xxx || new_expr.len() != range_plan.schema().fields().len();
```

I guess `xxx` should be something like `not_in_range_plan_schema`.

Or consider this:
```rust
let all_in_range_schema = new_expr.iter().all(|expr| {
    if let Expr::Column(column) = expr {
        range_plan.schema().has_column(column)
    } else {
        false
    }
});
let need_project = !all_in_range_schema || new_expr.len() != range_plan.schema().fields().len();
```